### PR TITLE
Feat/response

### PR DESCRIPTION
**Pull request by fabioz**
Although this is in draft, the main idea should be there already, so, checking for feedback on what you think about it @cmin764 @vlad-sema4

### CODE REVIEW

File: action_server/src/sema4ai/action_server/_actions_run.py [hunk 1]
```diff
@@ -222,34 +230,49 @@ def _run_action_in_thread(
...
+                    raise RuntimeError(
+                        "It was not possible to collect the contents of the "
+                        "result (json not created)."
+                    )
+
+                try:
+                    result_contents = json.loads(run_result_str)
+                except Exception:
+                    raise RuntimeError(
+                        f"Error loading the contents of {run_result_str} as json."
```
Comment (cmin764) [inline action_server/src/sema4ai/action_server/_actions_run.py:245]:
```suggestion
                        f"Error loading the contents of {run_result_str!r} as json."
```
If you want it repr-ed.

File: action_server/src/sema4ai/action_server/_actions_run.py [hunk 2]
```diff
@@ -222,34 +230,49 @@ def _run_action_in_thread(
...
+                    )
+
+                ret = result_contents.get("result")
+                result_str: str
+                if ret is None:
+                    result_str = "None"
```
Comment (cmin764) [inline action_server/src/sema4ai/action_server/_actions_run.py]:
```suggestion
                    result_str = "null"
```
if we'd like to inline it with JSON serialized value of Python's `None`. (but your preference here)

File: action_server/src/sema4ai/action_server/_actions_run.py [hunk 3]
```diff
@@ -222,34 +230,49 @@ def _run_action_in_thread(
...
+                else:
+                    result_str = json.dumps(ret, indent=4)
```
Comment (cmin764) [inline action_server/src/sema4ai/action_server/_actions_run.py]:
```suggestion
                result_str = json.dumps(result_contents.get("result"), indent=4)
```
isn't this simpler and more clean?

Example:
```python
>>> json.dumps({}.get("x"))
'null'
```

File: actions/pyproject.toml [hunk 4]
```diff
@@ -17,14 +17,12 @@ robocorp-log = ">=2.4,<3"
 robocorp-truststore= ">=0.8.0"
 psutil = "^5.0"
 docstring_parser_fork = "^0.0.5"
+pydantic = "^2.0" # Required for the `Response[T]`
```
Comment (cmin764) [inline actions/pyproject.toml:20]: Left a Slack [thread](https://sema4ai.slack.com/archives/C06CYLQ7S4R/p1716448556918039?thread_ts=1716382901.166559&cid=C06CYLQ7S4R).

I might be off-topic, just an idea.

File: actions/src/sema4ai/actions/_commands.py [hunk 5]
```diff
@@ -536,6 +579,14 @@ def run(
...
                 else:
                     retcode = int(any(action.failed for action in actions))
+                    # retcode = 0
+                    # for action in actions:
+                    #     if action.failed:
+                    #         if action.result is None:
+                    #             # This is the case where an ActionError is raised
+                    #             # (so, we have an exception but also a proper result).
+                    #             retcode = 1
+                    #             break
```
Comment (cmin764) [inline actions/src/sema4ai/actions/_commands.py]: Does it help to have dead code here? (or is this more like a reference of a potential upcoming TODO)

**Review by cmin764 — APPROVED** [review]
LGTM, left mostly optional nitpicks.

File: actions/src/sema4ai/actions/_commands.py (hunk 5 above)
Comment (fabioz) [reply actions/src/sema4ai/actions/_commands.py]: No, mostly leftovers from some experimentation. I'll definitely remove before the final version.

File: action_server/src/sema4ai/action_server/_actions_run.py (hunk 3 above)
Comment (fabioz) [reply action_server/src/sema4ai/action_server/_actions_run.py]: I was thinking about that too (it breaks tests, which is why I haven't done it, but I may end up doing that, just need to think a bit more about it).

File: action_server/src/sema4ai/action_server/_actions_run.py (hunk 1 above)
Comment (fabioz) [reply action_server/src/sema4ai/action_server/_actions_run.py:245]: Given that it's an indented json I think it's better not to do that (so messages may be more well formatted to show in the terminal).

File: action_server/src/sema4ai/action_server/_actions_run.py (hunk 2 above)
Comment (fabioz) [reply action_server/src/sema4ai/action_server/_actions_run.py]: Changed (but with `None` still being added to the error message)

**Pull request merged**